### ISSUE TITLE
practical methods for training LSTM using large corpus of data with keras

### ISSUE THREAD

**Issue by MrWanter** [fchollet/keras#5264]
I'm training a large LSTM network using around 10 million sequences of 6-12 length characters with Keras, I'm using the one-hot representation for input and output, with a large learning rate, it doesn't converge, but with a very small learning rate, it converges and the leaning curve looks fine, but got a validation result worse than large learning rate, can anyone suggest me a good way to find good hyperparams and a way to decide if I have got the best result the model can get?
thanks very much!

**Comment by Ori226** [comment]
Just an idea of for general LSTM debugging techniques :
* start with non-LSTM network on a fixed length sequence. Even MLP will probably provide better than average result.
* Start with the minimal length sequence for train, validation+ test on some LSTM network and see how it behave. From my experience, it is a good starting point.
* You mentioned you only succeed to converge with very small learning rate. Try build the network such that it will contain much fewer parameter. For example, you can create a first layer as temporal convolution in order to make the input size for the LSTM smaller. Only after you are able to achieve good results with the "simpler" network, start to increase the network complexity gradually. You can post your network architecture here if you want.


Good luck.

**Comment by MrWanter** [comment]
@Ori226  Thank you, I first trained the network on small set of sequences, around 20000 sequences, with rmsprop of learning rate 0.006   with the following network:
```
model = Sequential()
model.add(Masking(mask_value=0., input_shape=(None, len(chars))))
model.add(LSTM(150, return_sequences=True))
model.add(LSTM(150, return_sequences=True))
model.add(TimeDistributed(Dense(len(chars))))
model.add(Activation('softmax'))
```
it converges very fast and got a good result, but when I trained the network on the whole dataset of around 14 million sequences, with rmsprop of learning rate 0.0001 with the following network:

```
model = Sequential()
model.add(Masking(mask_value=0., input_shape=(None, len(chars))))
model.add(LSTM(4000, return_sequences=True))
model.add(LSTM(4000, return_sequences=True))
model.add(TimeDistributed(Dense(len(chars))))
model.add(Activation('softmax'))
```
the training loss does not decrease after only one epoch

**Comment by Ori226** [comment]
I think the first LSTM layer with 4000 is huge. Use model.summary() to see the number of parameters.  Have you tried the full dataset with the LSTM of 150 like the first example? BTW, can you explain what you are trying to do? Predict character?

**Comment by MrWanter** [comment]
I'm trying to train LSTM to learn distribution of large datasets of short sequences, like simple word or password,  just as character-level language model,  for choosing the network size, I set the size of parameters be around the same with the amount of chars in the datasets(as suggested by Karpathy [char-rnn](https://github.com/karpathy/char-rnn)), for example, for 20000 sequences, each is around 8-12 length, so there is around 200000 characters, then set the LSTM to be two layers of 150 neurons, which has approximately 300000 parameters (or two layers of 100 neurons, which has approximately 200000 parameters, this has a similar performance with 150 neurons), then for whole dataset of around 14 million sequences, there is around 140 million characters, so I set the network to be two layers of 4000 neurons, which has around 190 million parameters (I have tested with model.summary function).
I have not tried the full datasets with the original two layers of 150 neurons as it seems that will have the problem of heavy underfit

**Comment by MrWanter** [comment]
@Ori226 I'm trying to train LSTM to learn distribution of large datasets of short sequences, like simple word or password,  just as character-level language model,  for choosing the network size, I set the size of parameters be around the same with the amount of chars in the datasets(as suggested by Karpathy [char-rnn](https://github.com/karpathy/char-rnn)), for example, for 20000 sequences, each is around 8-12 length, so there is around 200000 characters, then set the LSTM to be two layers of 150 neurons, which has approximately 300000 parameters (or two layers of 100 neurons, which has approximately 200000 parameters, this has a similar performance with 150 neurons), then for whole dataset of around 14 million sequences, there is around 140 million characters, so I set the network to be two layers of 4000 neurons, which has around 190 million parameters (I have tested with model.summary function).
I have not tried the full datasets with the original two layers of 150 neurons as it seems that will have the problem of heavy underfit

**Comment by Ori226** [comment]
@MrWanter I read Karpathy's blog. I still suggest you to try with a much smaller LSTM layer. Maybe the problem is that due to the big LSTM layer, the training become very unstable (this is something that happened to me). My other suggestion is just general development suggestion - if something work in A and doesn't work in B, move gradually and  slowly from A to B and check the A-B intermediate results. I will be happy to know what you find.

**Comment by MrWanter** [comment]
@Ori226 Thank you, I'll try to slowly move from small dataset to large one and see if I can find some problem.

**Comment by NikhilKothari** [comment]
I am facing the same issue. I'm trying to predict next character using lstm. Everything works out perfectly fine for small dataset. But as training data explodes, training accuracy gets stuck at only 0.75. I've tried multiple similar architectures and hyper parameters as above but nothing seems to work. Please, oh please help!
